### PR TITLE
[DUCKDB] Native duckdb lance reader

### PR DESCRIPTION
**Pull request by eddyxu**
Support `SELECT * from lance_scan('s3://URI')` directly via `duckdb` extension.

### CODE REVIEW

**Comment by eddyxu** [comment]
Screenshot 2022-12-03 at 10 42 45 PM
Screenshot 2022-12-03 at 11 20 51 PM

**Comment by eddyxu** [comment]
Filed https://github.com/eto-ai/lance/issues/348 to track the optimization for zero-copy arrow=>ducdkb conversion.

File: integration/duckdb/src/lance/duckdb/lance_reader.cc [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+  return state;
+}
+
+/// Convert numeric array to duckdb vector.
+template <ArrowType T>
+void ToVector(const std::shared_ptr<::arrow::Array> &arr, ::duckdb::Vector *out) {
+  // TODO: dynamic_pointer_cast does not work here, IDK why.
+  auto array = std::static_pointer_cast<typename ::arrow::TypeTraits<T>::ArrayType>(arr);
+  assert(array != nullptr);
+  // TODO: How to use zero copy to move data from arrow to duckdb.
```
Comment (changhiskhan) [inline integration/duckdb/src/lance/duckdb/lance_reader.cc:104]: The duckdb-pyarrow integration is zero-copy right? i forgot how but vaguely remember seeing code that uses the Arrow C structs to dress up arrow memory buffer as duckdb vector. not a blocker for this tho i think

File: integration/duckdb/src/lance/duckdb/lance_reader.cc [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+  auto array = std::static_pointer_cast<::arrow::StringArray>(arr);
+  assert(array != nullptr);
+  // TODO: How to use zero copy to move data from arrow to duckdb.
+  for (int i = 0; i < array->length(); ++i) {
+    out->SetValue(i, std::string(array->Value(i)));
+  }
+  out->SetVectorType(::duckdb::VectorType::FLAT_VECTOR);
+}
+
+/// Convert a String array into duckdb vector.
```
Comment (changhiskhan) [inline integration/duckdb/src/lance/duckdb/lance_reader.cc:124]: Binary array?

File: integration/duckdb/src/lance/duckdb/lance_reader.cc [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+void ToVector<::arrow::DictionaryType>(const std::shared_ptr<::arrow::Array> &arr,
+                                       ::duckdb::Vector *out) {
+  auto array = std::static_pointer_cast<::arrow::DictionaryArray>(arr);
+  // TODO: zero copy
+  out->SetVectorType(::duckdb::VectorType::FLAT_VECTOR);
+  auto dict_arr = std::dynamic_pointer_cast<::arrow::StringArray>(array->dictionary());
+  auto indices_arr = std::static_pointer_cast<::arrow::Int8Array>(array->indices());
+  for (int i = 0; i < indices_arr->length(); ++i) {
+    auto idx = indices_arr->Value(i);
+    out->SetValue(i, std::string(dict_arr->Value(idx)));
```
Comment (changhiskhan) [inline integration/duckdb/src/lance/duckdb/lance_reader.cc:148]: i wonder whether this fixes the duckdb-arrow issue where list of dictionary querying doesn't work

**Review by changhiskhan — APPROVED** [review]
is it possible to add some unit tests?
perf optimization can happen separately

File: integration/duckdb/src/lance/duckdb/lance_reader.cc (hunk 1 above)
Comment (eddyxu) [reply integration/duckdb/src/lance/duckdb/lance_reader.cc:104]: I checked a few places, i.e.

* `pythonpkg` https://github.com/duckdb/duckdb/blob/master/tools/pythonpkg/src/include/duckdb_python/array_wrapper.hpp
* https://github.com/duckdb/duckdb/blob/master/src/include/duckdb/common/arrow/arrow_appender.hpp

Seems no code that convert from arrow => duckdb, it only has the reversed conversion (i..e, duckdb => arrow). 

I can look into it more later. Filed https://github.com/eto-ai/lance/issues/348 to track the work.

File: integration/duckdb/src/lance/duckdb/lance_reader.cc (hunk 3 above)
Comment (eddyxu) [reply integration/duckdb/src/lance/duckdb/lance_reader.cc:148]: Yea probally, we can actually make (a hardcoded set of ) user defined types work via our plugins tho.

File: integration/duckdb/src/lance/duckdb/lance_reader.cc (hunk 2 above)
Comment (eddyxu) [reply integration/duckdb/src/lance/duckdb/lance_reader.cc]: Good catch. fixed.

**Pull request merged**